### ISSUE TITLE
Increase `maxSignatureContextCount` limit (and/or make it configurable) for ParamSpec

### ISSUE THREAD

**Issue by bram-tv** [microsoft/pyright#6781]
(Preamble: I suppose this could also be considered an enhancement and not a bug)

## TLDR

The limit for `maxSignatureContextCount` is currently set to 64 but we're running into that limit.

## Use case

We're trying to wrap the `boto3.client()` method using `ParamSpec` but this runs into the `maxSignatureContextCount` limit.
The `boto3.client()` methods contains a `service_name` parameter and based on that a specific client is returned.
Boto3 itself doesn't contain type-hints but these are available in the boto3-stubs package.

For `boto3.client` there are (currently) 375 overloads defined.
One example (but all follow the same pattern):
```
@overload
def client(
    service_name: Literal["s3"],
    region_name: Optional[str] = ...,
    api_version: Optional[str] = ...,
    use_ssl: Optional[bool] = ...,
    verify: Union[bool, str, None] = ...,
    endpoint_url: Optional[str] = ...,
    aws_access_key_id: Optional[str] = ...,
    aws_secret_access_key: Optional[str] = ...,
    aws_session_token: Optional[str] = ...,
    config: Optional[Config] = ...,
) -> S3Client: ...
```

## Example

requirements.txt:
```
boto3
boto3-stubs[essential,s3,dynamodb]
pyright
typing_extensions
```

test.py:
```python
from typing_extensions import ParamSpec
from typing import TypeVar, Callable

import boto3


P = ParamSpec("P")
R = TypeVar("R")


def create_method(f: Callable[P, R]) -> Callable[P, R]:
    def wrapper(*args: P.args, **kwargs: P.kwargs) -> R:
        return f(*args, **kwargs)

    return wrapper


client2 = create_method(boto3.client)

x1 = boto3.client("s3")
reveal_type(x1)

x2 = boto3.client("dynamodb")
reveal_type(x2)

x3 = client2("s3")
reveal_type(x3)

x4 = client2("dynamodb")
reveal_type(x4)
```

Running with `pyright` as is (with `maxSignatureContextCount` set to 64):
```
.../test.py
  .../test.py:21:13 - information: Type of "x1" is "S3Client"
  .../test.py:24:13 - information: Type of "x2" is "DynamoDBClient"
  .../test.py:27:13 - information: Type of "x3" is "Unknown"
  .../test.py:30:13 - information: Type of "x4" is "Unknown"
0 errors, 0 warnings, 4 informations
```

When calling `boto3.client(...)` it was able to determine the correct return type but when using `client2(...)` (which simply wraps `boto3.client()`) it no longer was.

Increasing the `maxSignatureContextCount` limit to 376:
```
.../test.py
  .../test.py:21:13 - information: Type of "x1" is "S3Client"
  .../test.py:24:13 - information: Type of "x2" is "DynamoDBClient"
  .../test.py:27:13 - information: Type of "x3" is "S3Client"
  .../test.py:30:13 - information: Type of "x4" is "DynamoDBClient"
0 errors, 0 warnings, 4 informations
```

## A self-containing example

<details>
  <summary>Code</summary>

```python
from typing_extensions import ParamSpec
from typing import Any, TypeVar, Callable, Literal, overload


class Foo1: ...
class Foo2: ...
class Foo3: ...
class Foo4: ...
class Foo5: ...
class Foo6: ...
class Foo7: ...
class Foo8: ...
class Foo9: ...
class Foo10: ...
class Foo11: ...
class Foo12: ...
class Foo13: ...
class Foo14: ...
class Foo15: ...
class Foo16: ...
class Foo17: ...
class Foo18: ...
class Foo19: ...
class Foo20: ...
class Foo21: ...
class Foo22: ...
class Foo23: ...
class Foo24: ...
class Foo25: ...
class Foo26: ...
class Foo27: ...
class Foo28: ...
class Foo29: ...
class Foo30: ...
class Foo31: ...
class Foo32: ...
class Foo33: ...
class Foo34: ...
class Foo35: ...
class Foo36: ...
class Foo37: ...
class Foo38: ...
class Foo39: ...
class Foo40: ...
class Foo41: ...
class Foo42: ...
class Foo43: ...
class Foo44: ...
class Foo45: ...
class Foo46: ...
class Foo47: ...
class Foo48: ...
class Foo49: ...
class Foo50: ...
class Foo51: ...
class Foo52: ...
class Foo53: ...
class Foo54: ...
class Foo55: ...
class Foo56: ...
class Foo57: ...
class Foo58: ...
class Foo59: ...
class Foo60: ...
class Foo61: ...
class Foo62: ...
class Foo63: ...
class Foo64: ...


@overload
def foo(service_name: Literal["foo1"]) -> Foo1: ...
@overload
def foo(service_name: Literal["foo2"]) -> Foo2: ...
@overload
def foo(service_name: Literal["foo3"]) -> Foo3: ...
@overload
def foo(service_name: Literal["foo4"]) -> Foo4: ...
@overload
def foo(service_name: Literal["foo5"]) -> Foo5: ...
@overload
def foo(service_name: Literal["foo6"]) -> Foo6: ...
@overload
def foo(service_name: Literal["foo7"]) -> Foo7: ...
@overload
def foo(service_name: Literal["foo8"]) -> Foo8: ...
@overload
def foo(service_name: Literal["foo9"]) -> Foo9: ...
@overload
def foo(service_name: Literal["foo10"]) -> Foo10: ...
@overload
def foo(service_name: Literal["foo11"]) -> Foo11: ...
@overload
def foo(service_name: Literal["foo12"]) -> Foo12: ...
@overload
def foo(service_name: Literal["foo13"]) -> Foo13: ...
@overload
def foo(service_name: Literal["foo14"]) -> Foo14: ...
@overload
def foo(service_name: Literal["foo15"]) -> Foo15: ...
@overload
def foo(service_name: Literal["foo16"]) -> Foo16: ...
@overload
def foo(service_name: Literal["foo17"]) -> Foo17: ...
@overload
def foo(service_name: Literal["foo18"]) -> Foo18: ...
@overload
def foo(service_name: Literal["foo19"]) -> Foo19: ...
@overload
def foo(service_name: Literal["foo20"]) -> Foo20: ...
@overload
def foo(service_name: Literal["foo21"]) -> Foo21: ...
@overload
def foo(service_name: Literal["foo22"]) -> Foo22: ...
@overload
def foo(service_name: Literal["foo23"]) -> Foo23: ...
@overload
def foo(service_name: Literal["foo24"]) -> Foo24: ...
@overload
def foo(service_name: Literal["foo25"]) -> Foo25: ...
@overload
def foo(service_name: Literal["foo26"]) -> Foo26: ...
@overload
def foo(service_name: Literal["foo27"]) -> Foo27: ...
@overload
def foo(service_name: Literal["foo28"]) -> Foo28: ...
@overload
def foo(service_name: Literal["foo29"]) -> Foo29: ...
@overload
def foo(service_name: Literal["foo30"]) -> Foo30: ...
@overload
def foo(service_name: Literal["foo31"]) -> Foo31: ...
@overload
def foo(service_name: Literal["foo32"]) -> Foo32: ...
@overload
def foo(service_name: Literal["foo33"]) -> Foo33: ...
@overload
def foo(service_name: Literal["foo34"]) -> Foo34: ...
@overload
def foo(service_name: Literal["foo35"]) -> Foo35: ...
@overload
def foo(service_name: Literal["foo36"]) -> Foo36: ...
@overload
def foo(service_name: Literal["foo37"]) -> Foo37: ...
@overload
def foo(service_name: Literal["foo38"]) -> Foo38: ...
@overload
def foo(service_name: Literal["foo39"]) -> Foo39: ...
@overload
def foo(service_name: Literal["foo40"]) -> Foo40: ...
@overload
def foo(service_name: Literal["foo41"]) -> Foo41: ...
@overload
def foo(service_name: Literal["foo42"]) -> Foo42: ...
@overload
def foo(service_name: Literal["foo43"]) -> Foo43: ...
@overload
def foo(service_name: Literal["foo44"]) -> Foo44: ...
@overload
def foo(service_name: Literal["foo45"]) -> Foo45: ...
@overload
def foo(service_name: Literal["foo46"]) -> Foo46: ...
@overload
def foo(service_name: Literal["foo47"]) -> Foo47: ...
@overload
def foo(service_name: Literal["foo48"]) -> Foo48: ...
@overload
def foo(service_name: Literal["foo49"]) -> Foo49: ...
@overload
def foo(service_name: Literal["foo50"]) -> Foo50: ...
@overload
def foo(service_name: Literal["foo51"]) -> Foo51: ...
@overload
def foo(service_name: Literal["foo52"]) -> Foo52: ...
@overload
def foo(service_name: Literal["foo53"]) -> Foo53: ...
@overload
def foo(service_name: Literal["foo54"]) -> Foo54: ...
@overload
def foo(service_name: Literal["foo55"]) -> Foo55: ...
@overload
def foo(service_name: Literal["foo56"]) -> Foo56: ...
@overload
def foo(service_name: Literal["foo57"]) -> Foo57: ...
@overload
def foo(service_name: Literal["foo58"]) -> Foo58: ...
@overload
def foo(service_name: Literal["foo59"]) -> Foo59: ...
@overload
def foo(service_name: Literal["foo60"]) -> Foo60: ...
@overload
def foo(service_name: Literal["foo61"]) -> Foo61: ...
@overload
def foo(service_name: Literal["foo62"]) -> Foo62: ...
@overload
def foo(service_name: Literal["foo63"]) -> Foo63: ...
@overload
def foo(service_name: Literal["foo64"]) -> Foo64: ...

def foo(service_name: str) -> Any: ...


P = ParamSpec("P")
R = TypeVar("R")


def create_method(f: Callable[P, R]) -> Callable[P, R]:
   def wrapper(*args: P.args, **kwargs: P.kwargs) -> R:
        return f(*args, **kwargs)
   return wrapper

bar = create_method(foo)

x1 = foo("foo1")
reveal_type(x1)

x2 = foo("foo63")
reveal_type(x2)

x3 = bar("foo1")
reveal_type(x3)

x4 = bar("foo63")
reveal_type(x4)
```

</details>

Running the code as-is:
```
.../test2.py
  .../test2.py:215:13 - information: Type of "x1" is "Foo1"
  .../test2.py:218:13 - information: Type of "x2" is "Foo63"
  .../test2.py:221:13 - information: Type of "x3" is "Unknown"
  .../test2.py:224:13 - information: Type of "x4" is "Unknown"
0 errors, 0 warnings, 4 informations
```

After removing the overload for `Literal["foo64"]`:
```
.../test2.py
  .../test2.py:215:13 - information: Type of "x1" is "Foo1"
  .../test2.py:218:13 - information: Type of "x2" is "Foo63"
  .../test2.py:221:13 - information: Type of "x3" is "Foo1"
  .../test2.py:224:13 - information: Type of "x4" is "Foo63"
0 errors, 0 warnings, 4 informations
```

**Comment by bschnurr** [comment]
Seems they are trying to map `service_name` with return type.  I dont think there is currently a better way to define this.  Maybe a new Pep is needed as well
 
```python
@overload
def client(
    service_name: Literal["application-autoscaling"],
    region_name: Optional[str] = ...,
    api_version: Optional[str] = ...,
    use_ssl: Optional[bool] = ...,
    verify: Union[bool, str, None] = ...,
    endpoint_url: Optional[str] = ...,
    aws_access_key_id: Optional[str] = ...,
    aws_secret_access_key: Optional[str] = ...,
    aws_session_token: Optional[str] = ...,
    config: Optional[Config] = ...,
) -> ApplicationAutoScalingClient: ...
@overload
def client(
    service_name: Literal["application-insights"],
    region_name: Optional[str] = ...,
    api_version: Optional[str] = ...,
    use_ssl: Optional[bool] = ...,
    verify: Union[bool, str, None] = ...,
    endpoint_url: Optional[str] = ...,
    aws_access_key_id: Optional[str] = ...,
    aws_secret_access_key: Optional[str] = ...,
    aws_session_token: Optional[str] = ...,
    config: Optional[Config] = ...,
) -> ApplicationInsightsClient: ...
```

**Comment by bschnurr** [comment]
seems related to this discussion. https://github.com/python/typing/discussions/1412

**Comment by erictraut** [comment]
Yeah, I can bump the limit. I didn't think we'd ever see a function with so many overloads in real code. Some overload behaviors (e.g. when the type checker looks for overlapping overloads) have O(n^2) behavior in both pyright and mypy, so any number of overloads over about a dozen will potentially result in poor analysis performance. The maintainers of `boto` and/or `boto-stubs` may want to look for a mechanism that scales better.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.342, which I just published. It will also be included in a future release of pylance.
